test/jdk/javax/swing/JRadioButton/ButtonGroupFocus/ButtonGroupFocusTest.java

```diff
@@ -28,26 +28,56 @@
  * @run main ButtonGroupFocusTest
  */
 
-import javax.swing.*;
-import java.awt.*;
+import java.awt.AWTEvent;
+import java.awt.Container;
+import java.awt.FlowLayout;
+import java.awt.KeyboardFocusManager;
+import java.awt.Rectangle;
+import java.awt.Robot;
+import java.awt.Toolkit;
+import java.awt.event.AWTEventListener;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
 import java.awt.event.KeyEvent;
+import java.awt.image.BufferedImage;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.io.File;
+import java.util.concurrent.CountDownLatch;
 
-public class ButtonGroupFocusTest {
+import javax.imageio.ImageIO;
+import javax.swing.ButtonGroup;
+import javax.swing.JFrame;
+import javax.swing.JRadioButton;
+import javax.swing.SwingUtilities;
+
+import static java.awt.KeyboardFocusManager.getCurrentKeyboardFocusManager;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class ButtonGroupFocusTest {
 
     private static JRadioButton button1;
     private static JRadioButton button2;
     private static JRadioButton button3;
     private static JRadioButton button4;
     private static JRadioButton button5;
-    private static Robot robot;
+
+    private static final CountDownLatch button2FocusLatch = new CountDownLatch(1);
+    private static final CountDownLatch button3FocusLatch = new CountDownLatch(1);
+    private static final CountDownLatch button4FocusLatch = new CountDownLatch(1);
+
+    private static final CountDownLatch button2FocusLatch2 = new CountDownLatch(2);
```

Comment:
Minor: Should the no. of  button2FocusLatch2 latches be 1? since we already have another button2FocusLatch 

```suggestion
    private static final CountDownLatch button2FocusLatch2 = new CountDownLatch(1);
```

Reply:
No, `button2FocusLatch` tracks that _Button 2_ gets the initial focus; `button2FocusLatch2` tracks that _Button 2_ gets focus for the second time in case of repeated `KEY_PRESS` event for the <kbd>Tab</kbd> key.

If `button2FocusLatch2` is released, the test fails with a clear error message — _Button 2_ received focus whereas _Button 4_ is expected to have it. If the test continues, it will still fail because _Button 3_ won't get focus but there will be no indication why.

Reply:
I understand that you need 2 separate CountDownLatches for Button 2, but my question was about the count of  `button2FocusLatch2` . Shouldn't this be 1 instead of 2 counts or do you mean we require 2 counts of button2FocusLatch2 to track focus twice after the initial one?

Reply:
No, it has to be 2. Focus listeners are added before the frame is displayed.

https://github.com/openjdk/jdk/blob/e7991216fb25a0eeb094cbeef74128c6770c051a/test/jdk/javax/swing/JRadioButton/ButtonGroupFocus/ButtonGroupFocusTest.java#L100-L104

If Button 2 receives focus **twice**, `button2FocusLatch2` will be released. If it had 1, then the latch would be released when Button 2 gets focus at the test start.

Reply:
Makes sense. Thanks for clarifying.